src/System.Console/src/System/Console.cs

```diff
@@ -45,6 +50,70 @@ public static TextReader In
             }
         }
 
+        public static Encoding InputEncoding
+        {
+            get
+            {
+                return Volatile.Read(ref _inputEncoding) ?? EnsureInitialized(ref _inputEncoding, () =>
+                {
+                    return ConsolePal.InputEncoding;
+                });
+            }
+            set
+            {
+                CheckNonNull(value, "value");
+                lock (InternalSyncObject)
+                {
+                    // Set the terminal console encoding.
+                    ConsolePal.SetConsoleInputEncoding(value);
+
+                    Volatile.Write(ref _inputEncoding, value);
+
+                    // We need to reinitialize Console.In in the next call to _in
+                    // This will discard the current StreamReader, potentially 
+                    // losing buffered data
+                    Volatile.Write(ref _in, null);
+                }
+            }
+        }
+
+        public static Encoding OutputEncoding
+        {
+            get
+            {
+                return Volatile.Read(ref _outputEncoding) ?? EnsureInitialized(ref _outputEncoding, () =>
```

Comment:
I tried checking for all accesses and made sure we route everything via the property rather than the direct calls to the private fields. I was trying to be consistent with the rest of the design which I believe was used for performance gains. Although I agree using volatile here would make life easier :)